src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/TaskDeadlineService.kt

```diff
@@ -0,0 +1,64 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.service
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.ApprovedPremisesApplicationEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.AssessmentEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.PlacementApplicationEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.PlacementRequestEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.util.toLocalDateTime
+import java.time.Duration
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import kotlin.time.Duration.Companion.hours
+import kotlin.time.toJavaDuration
+
+@Service
+class TaskDeadlineService(
+  private val workingDayCountService: WorkingDayCountService,
+) {
+  fun getDeadline(assessment: AssessmentEntity): OffsetDateTime? {
+    val application = assessment.application
+    return when {
+      application !is ApprovedPremisesApplicationEntity -> null
+      application.isEmergencyApplication == true -> emergencyAssessmentDueDateTime(assessment)
+      application.isShortNoticeApplication() == true -> addWorkingDays(assessment.createdAt, SHORT_NOTICE_ASSESSMENT_TIMEFRAME)
+      else -> addWorkingDays(assessment.createdAt, STANDARD_ASSESSMENT_TIMEFRAME)
+    }
+  }
+
+  fun getDeadline(placementRequest: PlacementRequestEntity): OffsetDateTime {
+    val application = placementRequest.application
+    return when {
+      application.isEmergencyApplication == true -> placementRequest.createdAt
+      application.isShortNoticeApplication() == true -> addWorkingDays(placementRequest.createdAt, SHORT_NOTICE_PLACEMENT_REQUEST_TIMEFRAME)
+      application.isEsapApplication == true -> placementRequest.createdAt
+      else -> addWorkingDays(placementRequest.createdAt, STANDARD_PLACEMENT_REQUEST_TIMEFRAME)
+    }
+  }
+
+  fun getDeadline(placementApplication: PlacementApplicationEntity): OffsetDateTime {
+    return addWorkingDays(placementApplication.createdAt, STANDARD_PLACEMENT_APPLICATION_TIMEFRAME)
+  }
+
+  fun addWorkingDays(date: OffsetDateTime, workingDays: Int): OffsetDateTime = workingDayCountService.addWorkingDays(date.toLocalDate(), workingDays).toLocalDateTime()
+
+  private fun emergencyAssessmentDueDateTime(assessment: AssessmentEntity): OffsetDateTime {
+    val duration = if (assessment.createdAt.hour < 13) {
+      2.hours.toJavaDuration()
+    } else {
+      val nextWorkingDay = workingDayCountService.nextWorkingDay(assessment.createdAt.toLocalDate())
+      val nextWorkingDayAt11am = nextWorkingDay.atTime(11, 0)
+      Duration.between(nextWorkingDay, nextWorkingDayAt11am)
+    }
+
+    return assessment.createdAt.plus(duration).toInstant().atOffset(ZoneOffset.UTC)
+  }
+
+  companion object {
```

Comment:
Just wondering if in the commit message we should be clear on whether we are saying "within x days" or "in no more than x days" e.g.

> If it is a short notice application, it is due in 2 days

If the application is submitted today Monday Feb 26th is its assessment due before the end of Weds 28th or is it due before the end of Thurs 29th? Maybe it's worth having some explicit examples like this so that everyone is clear on the rules?

Reply:
Good plan, I've fixed the tests a bit more to be more clear